core/kernel/panic.c

```diff
@@ -4,11 +4,138 @@
  * Copyright (c) 2014, STMicroelectronics International N.V.
  */
 
+#include <kernel/interrupt.h>
+#include <kernel/misc.h>
 #include <kernel/panic.h>
 #include <kernel/thread.h>
 #include <kernel/unwind.h>
 #include <trace.h>
 
+static_assert(CFG_HALT_CORES_ON_PANIC_SGI >= 8 &&
+	      CFG_HALT_CORES_ON_PANIC_SGI < 16);
+
+#if defined(ARM32) && !defined(CFG_WITH_ARM_TRUSTED_FW)
+#if TRACE_LEVEL >= TRACE_DEBUG
+
+#include <kernel/spinlock.h>
+#include <sm/sm.h>
+
+static const char *const dump_table[] = {
+	"usr_sp",
+	"usr_lr",
+	"irq_spsr",
+	"irq_sp",
+	"irq_lr",
+	"fiq_spsr",
+	"fiq_sp",
+	"fiq_lr",
+	"svc_spsr",
+	"svc_sp",
+	"svc_lr",
+	"abt_spsr",
+	"abt_sp",
+	"abt_lr",
+	"und_spsr",
+	"und_sp",
+	"und_lr",
+#ifdef CFG_SM_NO_CYCLE_COUNTING
+	"pmcr",
+#endif
+#ifdef CFG_FTRACE_SUPPORT
+	"cntkctl",
+	"pad"
+#endif
+};
+
+static void dump_core_registers(void)
+{
+	struct sm_nsec_ctx *sm_nsec_ctx = sm_get_nsec_ctx();
+	static unsigned int lock = SPINLOCK_UNLOCK;
+	uint32_t *reg = NULL;
+	size_t i = U(0);
+
+	cpu_spin_lock(&lock);
+
+	if (!sm_nsec_ctx) {
+		cpu_spin_unlock(&lock);
+		return;
+	}
+
+	DMSG("CPU : %zu\n", get_core_pos());
+
+	reg = (uint32_t *)&sm_nsec_ctx->ub_regs.usr_sp;
+
+	for (i = U(0); i < ARRAY_SIZE(dump_table); i++)
+		DMSG("%10s : %#8"PRIx32, dump_table[i], reg[i]);
```

Comment:
Why is dumping these registers meaningful? Anyway, it doesn't belong here in generic code.

Reply:
It is useful to get information on a non-secure call that caused a panic(). Of course, only one of the core dump will be relevant.

Do you have a suggestion on where to put this code?

Reply:
Has it ever been used? I mean, we're normally entering OP-TEE in the same way so you'll likely to see the same thing each time, except perhaps that you'll be able to see if you entered due to a secure interrupt instead of an SMC. However, that should be easier to see via the panic stack trace.
Perhaps I'm missing something, but this seems like something that hasn't been tried in practice.

Reply:
This is intended to help finding out where non-secure world is executing on some specific error cases trapped by OP-TEE: mainly when a piece of non-secure code tries to access an unauthorized resource and when a secure watchdog is about to expire. We configure the related firewall driver in OP-TEE to raise a secure interrupt which print some driver useful related info and then panic (e.g. in [_plat-stm32mp1/plat_tzc400.c_](https://github.com/OP-TEE/optee_os/blob/4.0.0/core/arch/arm/plat-stm32mp1/plat_tzc400.c#L26-L30)). The panic traps all CPUs that (with the proposed change here) print their non-secure context.

Reply:
Back to my question. Has it ever been used? Do you know that the information is useful? It seems that you're missing the non-secure PC for instance.

Reply:
indeed `mon_spsr` and `mon_lr` fields are missing!